metis-media-service/src/main/java/eu/europeana/metis/mediaprocessing/RdfDeserializerImpl.java

```diff
@@ -6,50 +6,134 @@
 import eu.europeana.metis.mediaprocessing.model.EnrichedRdf;
 import eu.europeana.metis.mediaprocessing.model.EnrichedRdfImpl;
 import eu.europeana.metis.mediaprocessing.model.RdfResourceEntry;
-import eu.europeana.metis.mediaprocessing.model.RdfWrapper;
 import eu.europeana.metis.mediaprocessing.model.UrlType;
 import java.io.ByteArrayInputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
 import java.util.Set;
 import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import javax.xml.XMLConstants;
+import javax.xml.namespace.NamespaceContext;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathConstants;
+import javax.xml.xpath.XPathExpression;
+import javax.xml.xpath.XPathExpressionException;
+import javax.xml.xpath.XPathFactory;
 import org.jibx.runtime.IUnmarshallingContext;
 import org.jibx.runtime.JiBXException;
+import org.w3c.dom.Document;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
 
 /**
- * This object implements RDF deserialization functionality.
+ * This implements RDF deserialization functionality. The code that obtains the individual resources
+ * does not assume that we can convert the record to an EDM internal format. Link checking must also
+ * run on EDM external. We therefore use XPath expressions to obtain the required data.
  */
 class RdfDeserializerImpl implements RdfDeserializer {
 
   private final IUnmarshallingContext context;
 
+  private final XPathExpression getObjectExpression;
```

Comment:
We should clearly mention that this class is not thread safe on the documentation because of the XPathExpression..

Reply:
On the other hand(Please check the other comments on this class as well, because we need to probably choose one or the other option):
This class is instantiated per record right?
If so maybe it's better to make the XPath, XPathExpressions static and synchronize the evaluate method in synchronized blocks where we would probably gain a lot of speed...
What do you think?

Reply:
The interface that this class implements (called `RdfDeserializer` mentions this fact in the documentation.

Reply:
This class may be instantiated per record, but it doesn't have to be. It was designed to be instantiated per worker thread, which would provide the same speedup. I think that perhaps the best way to make this better is to create a maintenance ticket to do what you suggest (not just here but also in other parts of the code).

Reply:
Actually, I could do on-demand evaluation of the xpath expressions that would probably solve some of the issues that you mentioned.

Reply:
Ok let's create a separate ticket to possibly improve that.
But if you are saying that it's design to be instantiated per worker then we should definitely use Saxon because the worker threads can be on the same JVM and the default implementation has some static fields which can cause an issue..
And/Or synchronize those..
We found an issue like that recently(I forgot the component, I think transformation..)

Reply:
I made some changes to make things load lazily and in a thread-safe way. Could you create the ticket for what's left to do?